orx-quadtree/src/main/kotlin/Quadtree.kt

```diff
@@ -0,0 +1,215 @@
+            }
+        }
+    }
+
+    /**
+     * Finds the nearest and neighbouring points within a radius
+     *
+     * @param element
+     * @param radius
+     * @return
+     */
+    fun nearest(element: T, radius: Double): QuadtreeQuery<T>? {
+        val point = mapper(element)
+
+        if (!bounds.contains(point)) return null
+
+        val r2 = radius * radius
+
+        val scaledBounds = Rectangle.fromCenter(point, radius * 2)
+        val intersected: List<Quadtree<T>> = intersect(scaledBounds) ?: return null
+
+        var minDist = Double.MAX_VALUE
+        val nearestObjects = mutableListOf<T>()
+        var nearestObject: T? = null
+
+        for (interNode in intersected) {
+            for (obj in interNode.objects) {
+                if (element == obj) continue
+                val p = mapper(obj)
+
+                val dist = p.squaredDistanceTo(point)
+
+                if (dist < r2) {
+                    nearestObjects.add(obj)
+
+                    if (dist < minDist) {
+                        minDist = dist
+                        nearestObject = obj
+                    }
+                }
+            }
+        }
+
+        if (nearestObject == null) return null
+
+        return QuadtreeQuery(nearestObject, nearestObjects, intersected)
+    }
+
+    /**
+     * Inserts the element in the appropriate node
+     *
+     * @param element
+     * @return
+     */
+    fun insert(element: T): Boolean {
+        // only* the root needs to check this
+        if (depth == 0) {
+            if (!bounds.contains(mapper(element))) return false
+        }
+
+        if ((objects.size < maxObjects && isLeaf)) {
+            objects.add(element)
+
+            return true
+        }
+
+        if (isLeaf) subdivide()
+
+        objects.add(element)
+
+        for (obj in objects) {
+            val p = mapper(obj)
+            val x = if (p.x > bounds.center.x) 1 else 0
+            val y = if (p.y > bounds.center.y) 1 else 0
+            val nodeIndex = x + y * 2
+
+            nodes[nodeIndex]?.insert(obj)
+        }
+
+        objects.clear()
+
+        return true
+    }
+
+    /**
+     * Finds which node the element is within (but not necessarily belonging to)
+     *
+     * @param element
+     * @return
+     */
+    fun findNode(element: T): Quadtree<T>? {
+        val v = mapper(element)
+
+        if (!bounds.contains(v)) return null
+
+        if (isLeaf) return this
+
+        for (node in nodes) {
+            node?.findNode(element)?.let { return it }
+        }
+
+        return null
+    }
+
+    /**
+     * Draw the quadtree using batching
+     *
+     * @param batchBuilder
+     */
+    fun batch(batchBuilder: RectangleBatchBuilder) {
+        batchBuilder.rectangle(bounds)
+
+        for (node in nodes) {
+            node?.batch(batchBuilder)
+        }
+    }
+
+    /**
+     * Draw the quadtree
+     *
+     * @param drawer
+     */
+    fun draw(drawer: Drawer) {
```

Comment:
@edwinRNDR do Compositions have the functionality to achieve this? In which case, could a function like this return a Composition instead, without passing around the Drawer?

Reply:
It could but it is not as fast as the batch builders are. I don't see a problem with passing around the Drawer (it is pretty much made to be safely passed around)